docs/design/persistent-storage.md

```diff
@@ -0,0 +1,220 @@
+
+
+
+#### Pod to specify storage requirements
+
+```
+
+	Pod.Spec.Volumes.Volume 
+			Name
+			VolumeSource -- to be bound with available storage from the pool
+			PersistenceRequirements
+				Mode				ReadWriteOnce | ReadOnlyMany
+				RetentionPolicy		DeleteOnPodExit | RetainOnPodExit
+				Resources			ResourcesList (size, iops, throughput)
+				Selector			Labels to identify the PersistentStorage backing implementation
+
+```
+
+`PersistenceRequirements` are the capabilities and performance characteristics a pod author can request and rely on for the entire lifecycle of the volume.  Requests for storage that do not match available storage will go unfulfilled.
+
+
+#### Storage discovery 
+
+An API and CLI is needed to view available and in-use storage by namespace.
+
+**Use Case**:  Susie needs to crunch some data and serve it afterwards.  A single pod to create data is sufficient, but she estimates needing 10 web front ends to serve the data afterwards.  For this use case, the data does not change once created.  Susie first looks to see what is available in the cluster.
+
+```
+$ cluster/kubectl.sh get storage
+
+NAME                LABELS              MODES               RESOURCES
+_pv_vol_01          name=my-data        RWO,ROX             size=5gb,iops=30
+_pv_vol_02                              RWO,ROX             size=5gb,iops=30
+_pv_vol_03          name=other-data     RWO,ROX             size=25gb,iops=30
+
+```
+
+#### Request persistent storage for a pod
+
+Susie is happy to see storage that can be mounted many times (ReadOnlyMany).  That satisfies her load balancer requirement.  Susie creates an pod image with her number crunching code and requests persistence for her volume.
+
+Susie's YAML looks something like this for the pod that creates her data:
+
+
+```yaml
+
+id: creator-of-data
+kind: Pod
+namespace: ns
+apiVersion: v1beta1
+desiredState:
+  manifest:
+    version: v1beta1
+    id: creator-of-data
+    containers:
+      - name: numbercruncher
+        image: ns/makesomdata:v1
+        command:
+          - /run.sh
+        volumeMounts:
+          - name: data
+            mountPath: /some_data
+    volumes:
+      - name: data
+        persistenceRequirements:
+        	mode: readonlymany
+        	retentionPolicy: retainonpodexit
+        	resources:
+        		- size: 5
+        	selector:
+        		- name: my-persistent-data
+labels:
+  name: creator-of-data
+
+```
+
+#### Behind the curtain
+
+`PersistentStorageBinder` is watching for new pods with `PersistenceRequirements`.  The binder compares the request to all available storage and binds it to the closest match.
+
+To bind:
+
+* Set PersistentStorage.Status.PodRef = pod that originally requested storage
+* Set PersistentStorage.Labels = PersistenceRequirements.Selector
```

Comment:
I'm not sure what the most elegant way to achieve this is.  After pod exit, a persistent volume is essentially orphaned.  It has no identity on its own.  Other pods in the namespace must be able to request it by selector.  I would prefer to use a unique (to the namespace) name.